src/libostree/ostree-repo-finder-avahi.c

```diff
@@ -0,0 +1,998 @@
+
+  avahi_address_snprint (address_string, sizeof (address_string), address);
+
+  switch (address->proto)
+    {
+    case AVAHI_PROTO_INET6:
+      if (IN6_IS_ADDR_LINKLOCAL (address->data.data) ||
+          IN6_IS_ADDR_LOOPBACK (address->data.data))
+        return g_strdup_printf ("%s%%%d", address_string, interface);
+      /* else fall through */
+    case AVAHI_PROTO_INET:
+    case AVAHI_PROTO_UNSPEC:
+    default:
+      return g_strdup (address_string);
+    }
+}
+
+static OstreeAvahiService *
+ostree_avahi_service_new (const gchar        *name,
+                          const gchar        *domain,
+                          const AvahiAddress *address,
+                          AvahiIfIndex        interface,
+                          guint16             port,
+                          AvahiStringList    *txt)
+{
+  g_autoptr(OstreeAvahiService) service = NULL;
+
+  g_return_val_if_fail (name != NULL, NULL);
+  g_return_val_if_fail (domain != NULL, NULL);
+  g_return_val_if_fail (address != NULL, NULL);
+  g_return_val_if_fail (port > 0, NULL);
+
+  service = g_new0 (OstreeAvahiService, 1);
+
+  service->name = g_strdup (name);
+  service->domain = g_strdup (domain);
+  service->address = address_to_string (address, interface);
+  service->port = port;
+  service->txt = avahi_string_list_copy (txt);
+
+  return g_steal_pointer (&service);
+}
+
+/* Reference: RFC 6763, §6. */
+static gboolean
+parse_txt_record (const guint8  *txt,
+                  gsize          txt_len,
+                  const gchar  **key,
+                  gsize         *key_len,
+                  const guint8 **value,
+                  gsize         *value_len)
+{
+  gsize i;
+
+  g_return_val_if_fail (key != NULL, FALSE);
+  g_return_val_if_fail (key_len != NULL, FALSE);
+  g_return_val_if_fail (value != NULL, FALSE);
+  g_return_val_if_fail (value_len != NULL, FALSE);
+
+  /* RFC 6763, §6.1. */
+  if (txt_len > 8900)
+    return FALSE;
+
+  *key = (const gchar *) txt;
+  *key_len = 0;
+  *value = NULL;
+  *value_len = 0;
+
+  for (i = 0; i < txt_len; i++)
+    {
+      if (txt[i] >= 0x20 && txt[i] <= 0x7e && txt[i] != '=')
+        {
+          /* Key character. */
+          *key_len = *key_len + 1;
+          continue;
+        }
+      else if (*key_len > 0 && txt[i] == '=')
+        {
+          /* Separator. */
+          *value = txt + (i + 1);
+          *value_len = txt_len - (i + 1);
+          return TRUE;
+        }
+      else
+        {
+          return FALSE;
+        }
+    }
+
+  /* The entire TXT record is the key; there is no ‘=’ or value. */
+  *value = NULL;
+  *value_len = 0;
+
+  return (*key_len > 0);
+}
+
+/* TODO: Docs. Return value is only valid as long as @txt is. Reference: RFC 6763, §6. */
+GHashTable *
+_ostree_txt_records_parse (AvahiStringList *txt)
+{
+  AvahiStringList *l;
+  g_autoptr(GHashTable) out = NULL;
+
+  out = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, (GDestroyNotify) g_bytes_unref);
```

Comment:
I’ll try and bear that in mind, although I personally prefer putting all declarations at the start of a block since it helps me visualise the state in a function more easily (and whether there’s too much state and some refactoring is needed).